src/scripts/createLocales.ts

```diff
@@ -0,0 +1,217 @@
+import fs from "fs"
+import path from "path"
+import typescript, { ScriptTarget } from "typescript"
+import walkdir from "walkdir"
+import eol from "eol"
+import deepMerge from "deepmerge"
+import _ from "lodash"
+import scanner from "i18next-scanner"
+import vfs from "vinyl-fs"
+
+import type { Scanner } from "i18next-scanner"
+import type File from "vinyl"
+
+import { supportedLanguages } from "../utils/languages"
+
+/**
+ * What does this script?
+ *
+ * - First, it grabs all the existing translations under `i18n/merged/{lang}/index.json`
+ *  (the resource file)
+ *
+ * - Then it collects all the translation keys used in each file
+ *    1. Scans all the files under the `src/` folder
+ *    2. Looks for the translation keys used by calling `t({key})` or by using `<Translation id="{key}">`
+ *
+ * - Finally, under `i18n/locales`, it creates new json files with just the
+ *   translations needed for each specific namespace
+ */
+
+const input = ["./src/**/*.{ts,tsx}"]
+const output = "./i18n"
+
+/**
+ * Creates a namespace given a filename.
+ *
+ * E.g. for `usr/www/src/pages/get-eth.tsx` it returns `src-pages-get-eth`
+ *
+ * Namespaces generation logic:
+ *
+ * 1. One namespace per page is created. E.g. `src-pages-get-eth` or `src-templates-static`
+ *
+ * 2. A namespace to group all the translations under the `components/` folder.
+ *    These are going to be shared between multiple pages
+ *
+ * 3. A `common` namespace for the rest of the translations that are not included
+ *    either in a page or the components folder
+ */
+function createNamespaceFromFilename(filename: string) {
+  const file = path.parse(filename)
+  const dir = file.dir.slice(filename.lastIndexOf("src/"))
+  const namespace = dir.replace(/\//g, "-") + "-" + file.name
+  return namespace
+}
+
+/**
+ * Custom transform function that is going to allocate each translation to the
+ * corresponding namespace
+ */
+function customTransform(
+  this: Scanner,
+  file: File,
+  enc: BufferEncoding,
+  done: () => void
+) {
+  const { base, ext } = path.parse(file.path)
+  const extensions = [".ts", ".tsx"]
+  const target = ScriptTarget.ES2018
+
+  if (extensions.includes(ext) && !base.includes(".d.ts")) {
+    const content = fs.readFileSync(file.path, enc)
+
+    const { outputText } = typescript.transpileModule(content, {
+      compilerOptions: { target },
+      fileName: path.basename(file.path),
+    })
+
+    const namespace = createNamespaceFromFilename(file.path)
+
+    const customHandler = (key: string, options: any) => {
+      const ns = file.path.includes("components") ? "components" : namespace
+
+      // if the key is not found in the current lng then we store the
+      // defaultLng value instead
```

Comment:
@pettinarip Curious if there is a reason we're using "Lng" without the "a" for language in this script... we don't do this anywhere else in the codebase.

Reply:
Good catch. Didn't realize that. I guess I got confused with the terminology used by the scanner's code 🤦🏼 Will fix that.